analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs

```diff
@@ -31,26 +31,31 @@ public abstract class PureAttributeOnVoidMethodBase<TSyntaxKind> : SonarDiagnost
         protected const string DiagnosticId = "S3603";
         private const string MessageFormat = "Remove the 'Pure' attribute or change the method to return a value.";
 
-        private readonly DiagnosticDescriptor rule;
-
         protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public DiagnosticDescriptor Rule { get; }
 
         protected PureAttributeOnVoidMethodBase() =>
-            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, Language.RspecResources);
+            Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, Language.RspecResources);
 
         protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSymbolAction(
                 c =>
                 {
-                    if (c.Symbol is IMethodSymbol { ReturnsVoid: true } methodSymbol
-                        && !methodSymbol.Parameters.Any(p => p.RefKind != RefKind.None)
-                        && methodSymbol.GetAttributes().FirstOrDefault(a => a.AttributeClass.Is(KnownType.System_Diagnostics_Contracts_PureAttribute)) is { } pureAttribute)
+                    if (InvalidPureDataAttributeUsage(c.Symbol) is { } pureAttribute)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, pureAttribute.ApplicationSyntaxReference.GetSyntax().GetLocation()));
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, pureAttribute.ApplicationSyntaxReference.GetSyntax().GetLocation()));
                     }
                 },
                 SymbolKind.Method);
```

Comment:
it's weird that this doesn't work for local functions.. because it's still an `IMethodSymbol` - did you check to see for what [SymbolKind](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Symbols/SymbolKind.cs) the local function symbol is returned?

Reply:
nevermind - it's an open issue https://github.com/dotnet/roslyn/issues/35770